tests/Microsoft.Identity.Test.Unit/CDT/CdtAuthenticationScheme.cs

```diff
@@ -0,0 +1,263 @@
+        {
+            var header = new JObject();
+            header[JsonWebTokenConstants.Algorithm] = _cdtCryptoProvider.CryptographicAlgorithm;
+            header[JsonWebTokenConstants.Type] = JasonWebTokenType;
+            header[CdtClaimTypes.Nonce] = nonce;
+
+            var body = new JObject
+            {
+                // Mandatory parameters
+                [CdtClaimTypes.Constraints] = _constraints
+            };
+
+            string cdtConstraintToken = CreateJWS(JsonHelper.JsonObjectToString(body), JsonHelper.JsonObjectToString(header));
+            return cdtConstraintToken;
+        }
+
+        /// <summary>
+        /// A key ID that uniquely describes a public / private key pair. While KeyID is not normally
+        /// strict, AAD support for Cdt requires that we use the base64 encoded JWK thumbprint, as described by 
+        /// https://tools.ietf.org/html/rfc7638
+        /// </summary>
+        private static byte[] ComputeThumbprint(string canonicalJwk)
+        {
+            using (SHA256 hash = SHA256.Create())
+            {
+                return hash.ComputeHash(Encoding.UTF8.GetBytes(canonicalJwk));
+            }
+        }
+
+        /// <summary>
+        /// Creates a JWS (json web signature) as per: https://tools.ietf.org/html/rfc7515
+        /// Format: header.payload.signed_payload
+        /// </summary>
+        private string CreateJWS(string payload, string header, bool signPayload = true)
+        {
+            StringBuilder sb = new StringBuilder();
+            sb.Append(Base64UrlHelpers.Encode(Encoding.UTF8.GetBytes(header)));
+            sb.Append('.');
+            sb.Append(Base64UrlHelpers.Encode(payload));
+            string headerAndPayload = sb.ToString();
+
+            if (signPayload)
+            {
+                sb.Append('.');
+                sb.Append(Base64UrlHelpers.Encode(_cdtCryptoProvider.Sign(Encoding.UTF8.GetBytes(headerAndPayload))));
+            }
+
+            return sb.ToString();
+        }
+    }
+
+    public static class CdtClaimTypes
+    {
+        #region JSON keys for Http request
+
+        /// <summary>
+        /// Access token with response cnf
+        /// 
+        /// </summary>
+        public const string Ticket = "t";
+
+        /// <summary>
+        /// Constraints specified by the client
+        /// 
+        /// </summary>
+        public const string ConstraintsToken = "c";
+
+        /// <summary>
+        /// Constraints specified by the client
+        /// 
+        /// </summary>
+        public const string Constraints = "constraints";
+
+        /// <summary>
+        /// Non-standard claim representing a nonce that protects against replay attacks.
+        /// </summary>
+        public const string Nonce = "nonce";
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public const string Type = "typ";
+
+        #endregion
+    }
+
+    //TODO: Add support for ECD keys
```

Comment:
+1 :)

Reply:
But I actually don't think you need to worry about this too much.

MISE uses Wilson internally, and Wilson I think has helper APIs for this. See https://stackoverflow.com/questions/65512150/how-to-generate-a-json-web-key-jwk-in-c-sharp

I would use Wilson in all these examples.

Reply:
(and Wilson will abstract the cert type from you)